### PR TITLE
Gaussian observation

### PR DESCRIPTION
**Pull request by gviejo**
This pull request adds support for Gaussian observation models to the codebase, allowing users to model neural responses as Gaussian random variables in addition to the previously available models. The changes include documentation updates, code modifications to register and instantiate the new model, and the implementation of the `GaussianObservations` class with all necessary methods.

### Gaussian observation model support

* Added the `GaussianObservations` class to `src/nemos/observation_models.py`, implementing negative log-likelihood, log-likelihood, sample generation, deviance calculation, and scale estimation for Gaussian-distributed data.
* Registered `GaussianObservations` in the module's public API, the available observation models list, and the observation model builder, enabling its selection and instantiation via string identifiers. [[1]](diffhunk://#diff-a54af43ecbcfb9918d26f51ad66649b387df8982febbdcabc47dc99ad46e9fcdR19) [[2]](diffhunk://#diff-d0e5ffd93cd70dc52daec386dd00018cd761ce156a89367f517e4ea1711b55caR6-R9) [[3]](diffhunk://#diff-d0e5ffd93cd70dc52daec386dd00018cd761ce156a89367f517e4ea1711b55caR48-R49)

### Documentation and interface updates

* Updated documentation and docstrings in `docs/api_reference.rst` and `src/nemos/glm/glm.py` to mention and describe the Gaussian model as an alternative to existing observation models. [[1]](diffhunk://#diff-bb750029c24c7ccbbcae105dd0969dbf8d6ab9e65b225165c4b601a58dfc9574R146) [[2]](diffhunk://#diff-66e2425553be1d5ec2c434815177377b1de98fc8ed11df75f9e7080357fd2aa7R52-R54) [[3]](diffhunk://#diff-66e2425553be1d5ec2c434815177377b1de98fc8ed11df75f9e7080357fd2aa7L101-R104) [[4]](diffhunk://#diff-66e2425553be1d5ec2c434815177377b1de98fc8ed11df75f9e7080357fd2aa7L190-R193)
* Corrected and clarified parameter descriptions in sampling and scale estimation methods for other observation models to improve accuracy and consistency. [[1]](diffhunk://#diff-a54af43ecbcfb9918d26f51ad66649b387df8982febbdcabc47dc99ad46e9fcdL738-R739) [[2]](diffhunk://#diff-a54af43ecbcfb9918d26f51ad66649b387df8982febbdcabc47dc99ad46e9fcdL816-R817) [[3]](diffhunk://#diff-a54af43ecbcfb9918d26f51ad66649b387df8982febbdcabc47dc99ad46e9fcdL961-R961)

These changes collectively enable users to fit GLMs with Gaussian-distributed neural activity and provide all required methods for likelihood computation, sampling, and model evaluation.

Minimal example

``` python
import numpy as np
import nemos as nmo
import pynapple as nap
import matplotlib.pyplot as plt
import jax.numpy as jnp
import statsmodels.api as sm

np.random.seed(9876789)

nsample = 50
sig = 0.5
x = np.linspace(0, 20, nsample)
X = np.column_stack((x, np.sin(x), (x - 5) ** 2, np.ones(nsample)))
beta = [0.5, 0.5, -0.02, 5.0]

y_true = np.dot(X, beta)
y = y_true + sig * np.random.normal(size=nsample)

# Fit OLS
glm = nmo.glm.GLM(observation_model="Gaussian", solver_name = "LBFGS", solver_kwargs={"tol": 1e-12})
glm.fit(X, y)
y_pred = glm.predict(X)
scale = glm.observation_model.estimate_scale(y, y_pred, y.shape[0] - X.shape[1])

# FIt using jax
jax_w, residuals, rank, singular_vals = jnp.linalg.lstsq(X, y, rcond=None)
jaxscale = residuals[0] / (y.shape[0] - X.shape[1])

# Fit using statsmodels for comparison
res = sm.OLS(y, X).fit()
pred_ols = res.get_prediction()
iv_l = pred_ols.summary_frame()["obs_ci_lower"]
iv_u = pred_ols.summary_frame()["obs_ci_upper"]


fig, ax = plt.subplots(figsize=(8, 6))

ax.plot(x, y, "o", label="data")
ax.plot(x, y_true, "b-", label="True")
ax.plot(x, y_pred, "r--.", label="nemos GLM")
ax.plot(x, X @ jax_w, "g--", label="JAX Lstsq")
ax.plot(x, res.fittedvalues, "k--", label="Statsmodels OLS")
ax.legend(loc="best")

plt.show()

```

### CODE REVIEW

**Comment by BalzaniEdoardo** [comment]
Hi @gviejo , there are a few things missing at the moment:

## Source Code:
Almost LGTM, minor additions:

- [ ] add an identity to `inverse_link_function_utils.py`, use that as default link in GaussianObservations instead of a lambda function. Having a function with a name helps with the repr of the class;
- [ ] in `nemos/glm/initialize_parameters.py`, add an identity func to the registry  of inverse funcs: INVERSE_FUNCS and INVERSE_FUNCS_BY_SIMPLE_NAME

## Tests

- [ ] add the observation model to the parametrization of the `TestGLMObservationModel` and `TestPopulationGLMObservationModel`,  note that the fixtures `gaussianGLM_model_instantiation` and `population_gaussianGLM_model_instantiation` must be defined in conftest.py
    ```python
    @pytest.mark.parametrize(
        "model_instantiation",
        [
            "poissonGLM_model_instantiation",
            "gammaGLM_model_instantiation",
            "bernoulliGLM_model_instantiation",
            "negativeBinomialGLM_model_instantiation",
            "gaussianGLM_model_instantiation",  # New, must be defined in conftest.py
        ],
    )
    class TestGLMObservationModel:
    ```

    ```python
    @pytest.mark.parametrize(
        "model_instantiation",
        [
            "population_poissonGLM_model_instantiation",
            "population_gammaGLM_model_instantiation",
            "population_bernoulliGLM_model_instantiation",
            "population_negativeBinomialGLM_model_instantiation",
            "population_gaussianGLM_model_instantiation",  # New, must be defined in conftest.py
        ],
    )
    class TestPopulationGLMObservationModel:
    ```

- [ ] add Gaussian cases to observation-model-specific fixtures in `TestGLMObservationModel` class (around line 1454-1623). Each fixture needs an `elif "gaussian" in model_instantiation:` branch:
    - `ll_scipy_stats` (line ~1458): use `sm.families.Gaussian().loglike(y, mean_firing, scale=scale) / norm` (similar to Gamma)
    - `sklearn_model` (line ~1498): return `LinearRegression(fit_intercept=True)` (standard OLS)
    - `dof_lasso_strength` (line ~1522): return a reasonable value like `0.5` (may need empirical tuning)
    - `dof_lasso_dof` (line ~1542): return expected DOF arrays (similar to other models, may need empirical tuning)
    - `obs_has_defaults` (line ~1574): return `False` (GaussianObservations has scale parameter like Gamma)
    - `model_repr` (line ~1594): return expected repr string with `GaussianObservations()` and `identity` link function

- [ ] add a `TestGaussianGLM` class (after `TestGammaGLM` around line ~2900) which should be similar to `TestGammaGLM`. It runs tests specific to the observation model. The main difference with Gamma is in the link functions - use identity (canonical), exp, and log:
    ```python
    @pytest.mark.parametrize("inv_link", [identity, jnp.exp])  # identity from inverse_link_function_utils
    @pytest.mark.parametrize("glm_type", ["", "population_"])
    @pytest.mark.parametrize("model_instantiation", ["gaussianGLM_model_instantiation"])
    class TestGaussianGLM:
        """
        Unit tests specific to Gaussian GLM.
        """

        @pytest.mark.solver_related
        def test_fit_glm(self, inv_link, request, glm_type, model_instantiation):
            # Same pattern as TestGammaGLM.test_fit_glm
            # Should verify scale_ != 1 after fitting

        def test_score_glm(self, inv_link, request, glm_type, model_instantiation):
            # Same pattern as TestGammaGLM.test_score_glm

        def test_simulate_glm(self, inv_link, request, glm_type, model_instantiation):
            # Same pattern as TestGammaGLM.test_simulate_glm
    ```

- [ ] add fixtures to `conftest.py` for Gaussian GLM instantiation (follow the pattern of `gammaGLM_model_instantiation` and `population_gammaGLM_model_instantiation`):
    - `gaussianGLM_model_instantiation`: simulate Gaussian data with identity link, use `y = rate + np.random.normal(0, scale, size=rate.shape)` where `rate = X @ w_true + b_true`
    - `gaussianGLM_model_instantiation_pytree`: pytree version (same pattern as gamma)
    - `population_gaussianGLM_model_instantiation`: population version with multiple neurons
    - `population_gaussianGLM_model_instantiation_pytree`: population pytree version

**Note:** GaussianObservations has a scale parameter (variance) like GammaObservations, so fixtures and tests should handle scale similar to how they handle it for Gamma.

File: src/nemos/glm/glm.py [hunk 1]
```diff
@@ -98,7 +101,7 @@ class GLM(BaseRegressor[ModelParams]):
     ----------
     observation_model :
         Observation model to use. The model describes the distribution of the neural activity.
-        Default is the Poisson model.
+        Default is the Poisson model. Alternatives are "Gamma", "Bernoulli", "NegativeBinomial" and "Gaussian".
```
Comment (BalzaniEdoardo) [inline src/nemos/glm/glm.py:104]: nice addition

File: src/nemos/observation_models.py [hunk 2]
```diff
@@ -1383,6 +1383,210 @@ def estimate_scale(
...
+    def __init__(
+        self,
+    ):
+        super().__init__()
+        self.scale = 1.0
+
+    @property
+    def default_inverse_link_function(self):
+        """Identity link function for Gaussian observations."""
+        return lambda x: x
```
Comment (BalzaniEdoardo) [inline src/nemos/observation_models.py]: define an `identity` in the `inverse_function_utils.py` and use that

File: src/nemos/observation_models.py [hunk 3]
```diff
@@ -1383,6 +1383,210 @@ def estimate_scale(
...
+    ) -> jnp.ndarray:
+        r"""Compute the Gaussian negative log-likelihood.
+
+        This computes the Gaussian negative log-likelihood of the predicted rates
+        for the observed neural activity up to a constant.
+
+        Parameters
+        ----------
+        y :
+            The target activity to compare against. Shape (n_time_bins, ), or (n_time_bins, n_neurons).
```
Comment (BalzaniEdoardo) [inline src/nemos/observation_models.py]:
```suggestion
            The target activity to compare against. Shape ``(n_time_bins, )``, or ``(n_time_bins, n_neurons)``.
```

File: src/nemos/observation_models.py [hunk 4]
```diff
@@ -1383,6 +1383,210 @@ def estimate_scale(
...
+        predicted_rate :
+            The predicted rate of the current model. Shape (n_time_bins, ) or (n_time_bins, n_neurons).
```
Comment (BalzaniEdoardo) [inline src/nemos/observation_models.py]:
```suggestion
            The predicted rate of the current model. Shape ``(n_time_bins, )`` or ``(n_time_bins, n_neurons)``.
```

File: src/nemos/observation_models.py [hunk 5]
```diff
@@ -1383,6 +1383,210 @@ def estimate_scale(
...
+    def log_likelihood(
+        self,
+        y: jnp.ndarray,
+        predicted_rate: jnp.ndarray,
+        scale: Union[float, jnp.ndarray] = 1.0,
+        aggregate_sample_scores: Callable = lambda x: jnp.sum(jnp.mean(x, axis=0)),
+    ):
+        r"""Compute the Gaussian negative log-likelihood.
+
+        This computes the Gaussian log-likelihood of the predicted rates
```
Comment (BalzaniEdoardo) [inline src/nemos/observation_models.py]:
```suggestion
        Compute the Gaussian log-likelihood of the predicted rates
```

File: src/nemos/observation_models.py [hunk 6]
```diff
@@ -1383,6 +1383,210 @@ def estimate_scale(
...
+        .. math::
+            \ell(\mu,\sigma^2) = -\frac{n}{2}\log(2\pi\sigma^2) - \frac{1}{2\sigma^2}\sum_{i=1}^{n} (y_i - \mu_i)^2.
+
+        where :math:`\mu` is the predicted mean, :math:`\sigma^2` is the variance (scale), and :math:`y` is
+        the observed data.
+
+        Parameters
+        ----------
+        y :
+            The target activity to compare against. Shape (n_time_bins, ), or (n_time_bins, n_neurons).
```
Comment (BalzaniEdoardo) [inline src/nemos/observation_models.py]:
```suggestion
            The target activity to compare against. Shape ``(n_time_bins, )``, or ``(n_time_bins, n_neurons)``.
```

File: src/nemos/observation_models.py [hunk 7]
```diff
@@ -1383,6 +1383,210 @@ def estimate_scale(
...
+        predicted_rate :
+            The predicted rate of the current model. Shape (n_time_bins, ) or (n_time_bins, n_neurons).
```
Comment (BalzaniEdoardo) [inline src/nemos/observation_models.py]:
```suggestion
            The predicted rate of the current model. Shape ``(n_time_bins, )`` or ``(n_time_bins, n_neurons)``.
```

File: src/nemos/observation_models.py [hunk 8]
```diff
@@ -1383,6 +1383,210 @@ def estimate_scale(
...
+        scale :
+            The scale parameter of the model.
+        aggregate_sample_scores :
+            Function that aggregates the log-likelihood of each sample.
+
+        Returns
+        -------
+        :
+            The Gaussian log-likelihood. Shape (1,).
```
Comment (BalzaniEdoardo) [inline src/nemos/observation_models.py]:
```suggestion
            The Gaussian log-likelihood. Shape ``(1,)``.
```

File: src/nemos/observation_models.py [hunk 9]
```diff
@@ -1383,6 +1383,210 @@ def estimate_scale(
...
+        self,
+        key: jax.Array,
+        predicted_rate: jnp.ndarray,
+        scale: Union[float, jnp.ndarray] = 1.0,
+    ) -> jnp.ndarray:
+        """
+        Sample from the Gaussian distribution.
+
+        This method generates random numbers from a Gaussian distribution based on the given
+        `predicted_rate` and `scale`.
```
Comment (BalzaniEdoardo) [inline src/nemos/observation_models.py]:
```suggestion
        ``predicted_rate`` and ``scale``.
```

File: src/nemos/observation_models.py [hunk 10]
```diff
@@ -1383,6 +1383,210 @@ def estimate_scale(
...
+        predicted_rate :
+            Expected rate (mean) of the Gaussian distribution. Shape (n_time_bins, ) or
+            (n_time_bins, n_neurons).
+        scale :
+            The scale parameter (variance) for the distribution.
+
+        Returns
+        -------
+        jnp.ndarray
+            Random numbers generated from the Gaussian distribution based on the `predicted_rate` and the `scale`.
```
Comment (BalzaniEdoardo) [inline src/nemos/observation_models.py]:
```suggestion
            Random numbers generated from the Gaussian distribution based on the ``predicted_rate`` and the ``scale``.
```

File: src/nemos/observation_models.py [hunk 11]
```diff
@@ -1383,6 +1383,210 @@ def estimate_scale(
...
+        neural_activity: jnp.ndarray,
+        predicted_rate: jnp.ndarray,
+        scale: Union[float, jnp.ndarray] = 1.0,
+    ) -> jnp.ndarray:
+        r"""Compute the residual deviance for a Gaussian model.
+
+        Parameters
+        ----------
+        neural_activity:
+            The spike count activity. Shape (n_time_bins, ) or (n_time_bins, n_neurons) for population models.
```
Comment (BalzaniEdoardo) [inline src/nemos/observation_models.py]: double ticks

File: src/nemos/observation_models.py [hunk 12]
```diff
@@ -1383,6 +1383,210 @@ def estimate_scale(
...
+        predicted_rate:
+            The predicted firing rates. Shape (n_time_bins, ) or (n_time_bins, n_neurons) for population models.
```
Comment (BalzaniEdoardo) [inline src/nemos/observation_models.py]: double ticks

File: tests/test_observation_models.py [hunk 13]
```diff
@@ -33,6 +33,8 @@ def observation_model_rate_and_samples(observation_model_string, shape=None):
     elif observation_model_string == "Gamma":
         theta = 3
         y = jax.random.gamma(jax.random.PRNGKey(123), rate / theta) * theta
+    elif observation_model_string == "Gaussian":
+        y = rate  # + jax.random.normal(jax.random.PRNGKey(123), shape=rate.shape)
```
Comment (BalzaniEdoardo) [inline tests/test_observation_models.py]: probably needs some random noise
